pool.go

```diff
@@ -298,7 +295,101 @@ func (p *pool) fetchWith(ctx context.Context, c cid.Cid, with string) (blk block
 	return
 }
 
-func (p *pool) fetchAndUpdate(ctx context.Context, node string, c cid.Cid, attempt int, transientErrs map[string]error) (blk blocks.Block, err error) {
+func (p *pool) fetchResourceWith(ctx context.Context, path string, cb DataCallback, with string) (err error) {
+	// wait for pool to be initialised
+	<-p.started
+
+	transientErrs := make(map[string]error)
+
+	left := p.config.MaxRetrievalAttempts
+	aff := with
+	if aff == "" {
+		aff = path
+	}
+
+	p.lk.RLock()
+	if left > len(p.endpoints) {
+		left = len(p.endpoints)
+	}
+
+	// if there are no endpoints in the consistent hashing ring, we submit a pool refresh request and fail this fetch.
+	if p.c == nil || p.c.Size() == 0 {
+		p.lk.RUnlock()
+		return ErrNoBackend
+	}
+	nodes, ok := p.c.GetNodes(aff, left)
+	p.lk.RUnlock()
+
+	// if there are no endpoints in the consistent hashing ring for the given cid, we submit a pool refresh request and fail this fetch.
+	if !ok || len(nodes) == 0 {
+		select {
+		case p.refresh <- struct{}{}:
+		default:
+		}
+
+		return ErrNoBackend
+	}
+
+	carFetchStart := time.Now()
+
+	pq := []string{path}
+	for i := 0; i < len(nodes); i++ {
+		err = p.fetchResourceAndUpdate(ctx, nodes[i], pq[0], i, cb, transientErrs)
+
+		if err == nil {
+			pq = pq[1:]
```

Comment:
could panic. We should check the length first ?

Reply:
pq[0] is referenced on line 337, there is no case at the end of the loop where pq will have length 0, right?